### DISCUSSION TITLE
Using opensnitch as router firewall on Raspberry Pi

### DISCUSSION THREAD

**Discussion by temp5mr** [evilsocket/opensnitch#456] [Q&A] [answered]
Hi. Please tell if it is possible to "catch" outgoing packets from wi-fi (wlan0) interface to outside (eth1).
I try use the Raspbery Pi 4 with **hostapd** and **isc-dhcp-server**, but it looks like **v1.4.0rc2** don't work with any packets except localhosted.
Sorry for a noob question but may be I should use another iptables rules instead the standard:

1) I use
    -P INPUT DROP
    -P FORWARD DROP
    -P OUTPUT ACCEPT
    -A INPUT -i lo -j ACCEPT
    -A INPUT -m state --state RELATED,ESTABLISHED -j ACCEPT
    -A INPUT -p tcp --dport 22 -i wlan0 -j ACCEPT
    -A INPUT -j DROP
    -A FORWARD -i eth1 -o wlan0 -m state --state RELATED,ESTABLISHED -j ACCEPT
    -A FORWARD -i wlan0 -o eth1 -j ACCEPT
    -A FORWARD -j DROP
    -N PORT-SCAN
    -A PORT-SCAN -p tcp --tcp-flags SYN,ACK,FIN,RST RST -m limit --limit 1/s -j RETURN
    -A PORT-SCAN -j DROP
    -t nat -A POSTROUTING -o eth1 -j MASQUERADE
2) then opensnitch says _firewall rules changed, reloading_
3) but nothing hapen when computer behind the router goes outside.

Is there something to try/test/learn... or not even think about it?

**Comment by gustavo-iniguez-goya** [comment, marked as answer]
Hi @temp5mr ,

Yes, it's possible to catch forwarded connections with the following rule:
`iptables -t mangle -I FORWARD -m conntrack --ctstate NEW -j NFQUEUE --queue-num 0 --queue-bypass`

However (as of today) we require the connections be initiated from the same machine where the daemon is installed. That means that we can intercept connections initiated from Virtual Machines or containers, but no connections that are traversing the box.

In this scenario the connections are traversing the box, originated from outside of the Pi, so even if they're intercepted we won't apply the rules on them because we're unable to get the PID who initiated them.

This is actually a limitation, because we can have rules that applies only to connections properties, like: drop everything to a.b.c.d. In this case we shouldn't require a PID to apply some rules.

**Reply by temp5mr** [reply]
Thank you. I've tryed. No way without PID((

Is it possible to add to opensnitch the "no pid" fake-process to manage the rules for outside-initialized connections?

**Reply by temp5mr** [reply]
May be try to use something that give us local PID - transparent proxy like squid, for example?

**Reply by gustavo-iniguez-goya** [reply]
We could split the rules set into 2 types of rules: network rules and application rules. 

Network rules would be applied without looking for the Inode/PID, saving a lot of CPU cycles. Besides supporting this scenario where we'd be able to block connections traversing the box, rules matching only properties of a connection would be much faster. 

Application rules would be those that require looking for the PID of the app, in order to match properties of it (path, name, etc).

But this is a quite intrusive change, and I don't think it'll be added soon.

**Comment by gustavo-iniguez-goya** [comment]
For future references:

We could return the connection object here:
https://github.com/evilsocket/opensnitch/blob/fbcab5a0c4419b9b2bca438ebb5b47cc54f2e3ae/daemon/conman/connection.go#L115

       return c, fmt.Errorf("no inodes found, applying default action.")

change this here: 
https://github.com/evilsocket/opensnitch/blob/fbcab5a0c4419b9b2bca438ebb5b47cc54f2e3ae/daemon/conman/connection.go#L45
```
	if con == nil {
            log.Debug("%s", err)
            return nil
        }
```

Check all the conn.Process for nil in operator.go, main.go, stats.go and connection.go , and insert a fw rule to redirect new forwarded connections (`-t mangle -I FORWARD -m conntrack --ctstate NEW -j NFQUEUE --queue-num 0 --queue-bypass`) to nfqueue 0.

We could then apply rules on connections, regardless if they're inbounds, outbunds or forwarded.
